gpu: Better screen support (and other things to pander to the panders)

GPUCode opened the pull request:
Same as the previous PR but trimmed to the essentials. ~~Fb rectangles with be in the next PR~~.

wheremyfoodat commented on src/core/services/gsp_gpu.cpp:
```diff
@@ -19,16 +21,14 @@ namespace ServiceCommands {
 }
 
 // Commands written to shared memory and processed by TriggerCmdReqQueue
-namespace GXCommands {
```
See previous

wheremyfoodat commented on src/core/PICA/regs.cpp:
```diff
@@ -162,7 +187,7 @@ void GPU::writeInternalReg(u32 index, u32 value, u32 mask) {
 			break;
 
-		// Restart immediate mode primitive drawing
+			// Restart immediate mode primitive drawing
```
Sir your Qt creator is broken

wheremyfoodat commented on include/renderer_gl/renderer_gl.hpp:
```diff
@@ -44,8 +44,8 @@ class RendererGL final : public Renderer {
 	float oldDepthOffset = 0.0;
 	bool oldDepthmapEnable = false;
 
-	SurfaceCache<DepthBuffer, 10, true> depthBufferCache;
```
Is there a particular use case that inspired going from 10 to 64?

GPUCode replied:
Had some crashes and bumping the number makes games more stable.

wheremyfoodat replied:
Try configuring CMake with -DDISABLE_PANIC_DEV=ON
This will make these games not crash and use a ring buffer instead

wheremyfoodat replied:
(Prolly need to delete CMakeCache first)

GPUCode replied:
Should I keep the bumped number or revert it back to 10?

wheremyfoodat replied:
I'll bump it down for now and consider more carefully how to adjust it in the guture should the need arise

wheremyfoodat commented on include/renderer_gl/surfaces.hpp:
```diff
@@ -29,12 +33,32 @@ struct ColourBuffer {
         // If this becomes a bottleneck, we can fix it semi-easily
         auto prevTexture = OpenGL::getTex2D();
-        texture.create(size.x(), size.y(), GL_RGBA8);
+        texture.create(size.x(), size.y(), internalFormat);
```
Won't this end up degrading image quality? I initially used GL_RGBA8 here always hoping that the extended precision would make for nicer textures.

GPUCode replied:
It's a question if you want to be more accurate to the hardware or better image quality. I haven't really tested if this affects image quality, though having the correct format makes texture downloads simpler. Also in vulkan if you have the correct format you can do reinterpretation with a simple image copy instead of shader in some cases.

wheremyfoodat replied:
Reverting temporarily

ArbestRiAgain commented:
Fixes the stretched bottom screen in 3D Land and many other games.
![image](https://github.com/wheremyfoodat/Panda3DS/assets/132712354/0d5c2265-120c-41d6-91c1-b65312a151c0)

wheremyfoodat commented on include/renderer_gl/opengl.hpp:
```diff
@@ -615,83 +615,4 @@ namespace OpenGL {
     }
 
-    // Abstraction for GLSL vectors
-    template <typename T, int size>
```
There was no way for you to know this lol but I'll bring these back. Reason is that this header is actually a standalone library several other projects use so I need to be able to drag and drop the latest Panda version to those projects. This means I can't make breaking ABI changes like this

GPUCode replied:
The header should be in externals then

wheremyfoodat replied:
True imma move it when I'm on the keeb again

wheremyfoodat commented:
Last commit fixes an "issue" where the compiler straight up does not optimize `gpu.readExternalRegs(Framebuffer0Addr)` and emits a whole load GPU object + call readExternalRegs + bounds check + read + return. This is actually impossible to optimize without aggressive LTO (and that is still an if).

Now it should be closer to a regular variable access instead.

I am also not sure if my relocating externalRegs for the reason I mentioned is 100% a good idea. My intuition says so, but I may reconsider in the future.

wheremyfoodat commented on include/renderer_gl/surfaces.hpp:
```diff
@@ -79,7 +113,7 @@ struct ColourBuffer {
     u32 location;
     PICA::DepthFmt format;
-    OpenGL::uvec2 size; // Implicitly set to the size of the framebuffer
+    Math::uvec2 size; // Implicitly set to the size of the framebuffer
```
Nowadays we have glm too btw for all vector/matrix/quaternion needs but the "old" vec implementation is fine here too since there's no math being really done.

(This is just an informational comment, nothing needs to change here)

wheremyfoodat commented on src/core/services/gsp_gpu.cpp:
```diff
@@ -261,6 +261,18 @@ void GPUService::setAxiConfigQoSMode(u32 messagePointer) {
 	mem.write32(messagePointer + 4, Result::Success);
 }
 
+void GPUService::setBufferSwap(u32 messagePointer) {
```
This would make games crash via err:f since it does not handle the IPC response header or the IPC result code. I am not sure what game to use to test it though :/

If there's none we know of atm, I will keep the code but add a Helpers::panic to revisit this when a game like that is found.

wheremyfoodat replied:
Fixed the IPC buffer for SetBufferSwap, still added a panic for when there's a known test case

GPUCode commented on src/core/renderer_gl/renderer_gl.cpp:
```diff
@@ -550,42 +579,64 @@ OpenGL::Texture RendererGL::getTexture(Texture& tex) {
-	// Displays are hard I really don't want to try implementing them because getting a fast solution is terrible
-	OpenGL::Texture& tex = framebuffer.has_value() ? framebuffer.value().get().texture : colourBufferCache[0].texture;
+	if (inputWidth != outputWidth) {
+		// Helpers::warn("Strided display transfer is not handled correctly!\n");
```
Why did you comment this out? Games may silently break due to that

wheremyfoodat replied:
FE: A has it trigger too much to the point of making it crawl.
I can lock it behind an

```cc
if constexpr (!Helpers::isUserBuild()) {}
```

GPUCode commented:
Majoras Mask is broken because of strided copy. I'll add it to this PR

wheremyfoodat commented:
OK, leave a message here when I can continue

GPUCode commented:
Added basic texture copy implementation that should work in the majority of cases. I've also labeled textures and added debug scopes for display transfer and texture copy so debugging in renderdoc is made easier
![εικόνα](https://github.com/wheremyfoodat/Panda3DS/assets/47210458/7b7ee2c2-7f92-4b2d-b97f-0ffca9ff6b41)

wheremyfoodat commented on src/core/renderer_gl/renderer_gl.cpp:
```diff
@@ -616,18 +615,88 @@ void RendererGL::displayTransfer(u32 inputAddr, u32 outputAddr, u32 inputSize, u
 	// Blit the framebuffers
-	srcFramebuffer.fbo.bind(OpenGL::ReadFramebuffer);
-	destFramebuffer.fbo.bind(OpenGL::DrawFramebuffer);
+	srcFramebuffer->fbo.bind(OpenGL::ReadFramebuffer);
```
Unchecked optional access?

GPUCode replied:
optional only matters if createIfNotFound is false, otherwise the function is guarantted to return a surface

wheremyfoodat commented:
Threw the format calls for OpenGL debug behind `#ifdef GPU_DEBUG_INFO` because they are impossible to optimize out otherwise.
I will look to see why NSMB2 is not a huge fan of the textureCopy implementation

GPUCode commented:
Latest commit fixes jitter in MK7 bottom screen

wheremyfoodat commented:
Resolving conflicts in a bit

wheremyfoodat commented:
Should be resolved

wheremyfoodat commented:
More conflicts
![256956941-cf3ef8ad-5b68-4885-a0e2-67651f20ee3a](https://github.com/wheremyfoodat/Panda3DS/assets/44909372/4290cbc2-b57b-49e3-a06d-d2dc139b33c4)

GPUCode commented:
Latest commit fixes Band of Thieves and 1000 Pokemon and also corrects the initial splash screen in Captain Toad
![εικόνα](https://github.com/wheremyfoodat/Panda3DS/assets/47210458/8e643922-87e5-40c3-851c-f198fb7563eb)

wheremyfoodat commented:
Omg
> Band of Thieves and 1000 Pokemon

Sir you're just making up games now

wheremyfoodat commented:
Any game to test v flip with?

GPUCode commented:
> Any game to test v flip with?

Don't remember any off the top of my head, I added it because it was quite trivial

wheremyfoodat commented on src/core/renderer_gl/renderer_gl.cpp:
```diff
@@ -473,6 +473,7 @@ void RendererGL::display() {
 	auto bottomScreen = colourBufferCache.findFromAddress(bottomScreenAddr);
 	screenFramebuffer.bind(OpenGL::DrawFramebuffer);
 
+	OpenGL::setClearColor(0.f);
```
I will add clear colour/depth/stencil to gl_state because might as well

wheremyfoodat commented:
![image](https://github.com/wheremyfoodat/Panda3DS/assets/44909372/46c83338-c260-4a2c-aea9-97e511cc41a0)
![image](https://github.com/wheremyfoodat/Panda3DS/assets/44909372/b48436d5-b60b-4265-856e-62582ed89ee4)
